conmon-rs/server/src/streams.rs

```diff
@@ -42,120 +36,90 @@ impl Streams {
-                Self::read_loop_single_stream(
-                    logger,
-                    Pipe::StdErr,
-                    message_tx,
-                    stop_rx,
-                    stderr.as_raw_fd(),
-                )
-                .await
+                Self::read_loop(stderr.as_raw_fd(), Pipe::StdErr, logger, message_tx).await
             });
         }
     }
 
-    async fn read_loop_single_stream(
-        logger: SharedContainerLog,
-        pipe: Pipe,
-        message_tx: mpsc::UnboundedSender<Message>,
-        mut stop_rx: broadcast::Receiver<()>,
+    pub async fn read_loop(
         fd: RawFd,
-    ) -> Result<()> {
-        trace!("Start reading from single fd: {:?}", fd);
-
-        let message_tx_clone = message_tx.clone();
-        let (thread_shutdown_tx, thread_shutdown_rx) = oneshot::channel();
-        task::spawn(async move {
-            Self::run_buffer_loop(logger, pipe, message_tx_clone, fd, thread_shutdown_rx).await
-        });
-
-        stop_rx
-            .recv()
-            .await
-            .context("unable to wait for stop channel")?;
-        debug!("Received IO stream stop signal");
-        thread_shutdown_tx
-            .send(())
-            .map_err(|_| format_err!("unable to send thread shutdown message"))?;
-        message_tx.send(Message::Done).context("send done message")
-    }
-
-    async fn run_buffer_loop(
-        logger: SharedContainerLog,
         pipe: Pipe,
+        logger: SharedContainerLog,
         message_tx: mpsc::UnboundedSender<Message>,
-        fd: RawFd,
-        mut thread_shutdown_rx: oneshot::Receiver<()>,
     ) -> Result<()> {
         let stream = unsafe { File::from_raw_fd(fd) };
-        let mut reader = BufReader::with_capacity(1024, stream);
+        let mut reader = BufReader::new(stream);
         let mut buf = vec![0; 1024];
         loop {
-            select! {
-                res = reader.read(&mut buf) => match res {
-                    Ok(n) if n > 0 => {
-                        debug!("Read {} bytes", n);
-                        let data = &buf[..n];
-
-                        let mut locked_logger = logger.write().await;
-                        locked_logger
-                            .write(pipe, data)
-                            .await
-                            .context("write to log file")?;
+            match reader.read(&mut buf).await {
+                Ok(n) if n > 0 => {
+                    debug!("Read {} bytes", n);
+                    let data = &buf[..n];
+
+                    let mut locked_logger = logger.write().await;
+                    locked_logger
+                        .write(pipe, data)
+                        .await
+                        .context("write to log file")?;
+
+                    message_tx
+                        .send(Message::Data(data.into()))
+                        .context("send data message")?;
+                }
+                Ok(n) if n == 0 => {
```

Comment:
In the non-terminal case we always get one read with `n == 0` before the actual data comes in. We could work around this with something like:
```diff
diff --git a/conmon-rs/server/src/streams.rs b/conmon-rs/server/src/streams.rs
index 9e3c28a..1cdb10a 100644
--- a/conmon-rs/server/src/streams.rs
+++ b/conmon-rs/server/src/streams.rs
@@ -73,6 +73,7 @@ impl Streams {
         let stream = unsafe { File::from_raw_fd(fd) };
         let mut reader = BufReader::new(stream);
         let mut buf = vec![0; 1024];
+        let mut already_read = false;
         loop {
             match reader.read(&mut buf).await {
                 Ok(n) if n > 0 => {
@@ -88,8 +89,10 @@ impl Streams {
                     message_tx
                         .send(Message::Data(data.into()))
                         .context("send data message")?;
+
+                    already_read = true;
                 }
-                Ok(n) if n == 0 => {
+                Ok(n) if n == 0 && already_read => {
                     debug!("No more to read");
 
                     message_tx
```

Reply:
do you know why that is?

Reply:
I have no idea unfortunately.

Reply:
do you think it was what @rphillips found where we sent Message::Done twice?